### ISSUE TITLE
Update to Solr 8 (Latest)

### ISSUE THREAD

**Issue by cdrini** [internetarchive/openlibrary#3317]
With the reindex flow largely complete (#1843), we can start looking at the next step, which is updating solr to the latest version. This is largely implemented in #2246 , but needs the last push to get it to production and to be up-to-date with #1843. This will allow:

- Near Real Time updating! New solr lets edits appear in virtually real time, since softCommits allow for data to become searchable significantly faster. Blocking #789 .
- Better built-in non-English character support! ~Blocking #714 , #178
- Partial document updates! We'd be able to update parts of a record without doing a full re-index of the document. Although out of scope of this issue, this would allow us long term to make solr-updater _siginficiantly_ faster. It also unblocks keeping counters like sponsorship status, or reading-log stats in solr. Blocking #3249 .

### Plan
- [x] Get Solr 8 working on local environment
    - [x] `make reindex-solr` works
    - [x] Searching via the UI works
    - [x] PR with cross-solr changes: #4283 
    - [x] Draft PR with solr 8 additions (Subset of @tfmorris ' #2246 )
        - Maybe a non-draft PR with a flag that lets people try using solr 8?
- [x] Try a full reindex with solr 8
    - [x] Switch solr_builder to use solr8
    - [x] Index works
    - [x] Index authors
    - [x] Switching staging to point to newly indexed solr, see what works/doesn't work
    - [x] Index orphans
    - [x] Index subjects: How will we import subjects? 😬 They're currently locked on solr 3.6 . ~Might need to do some sort of programatic import 😞~ Nope! Just built a fresh indexer :) (see comment below)
- [x] Test with solr-updater to catchup
- [x] Attach to testing.openlibrary.org
- [x] Fix bugs
    - [x] `text_general` converts everything to `multiValued`! author_name is now `multiValued` :/
    - [x] `title:` field not working with quotes sometimes
    - [x] search `all` failing for "The Mark of the Crown"
    - [x]  `key:(/works/OL450777W OR /works/OL362289W)` no longer works; need to do `key:("/works/OL450777W" OR "/works/OL362289W")`
- [x] Do another fresh reindex which will then go to prod!
- [x] Test on testing.openlibrary.org
- [x] Trial on one web node
- [x] Trial on both web nodes for a week
- [x] Fix any issues
    - [x] #5209 @finnless noting poor results for "foundation" in all search
- [x] Deploy!
- [ ] #4337 merged, Close this issue :)
- [x] Create issues for fixes/features to be done at a later date
    - [x] #5393 Re-enable english stopword detection for `text_en_splitting`
    - [x] #5343
    - [x] #5394 Delete old solr configs/etc.

### Proposal & Constraints
1. Pull out commits related to solr update from #2246 into separate PR
    - Replace custom solr docker image with latest official solr image
    - Use default schema.xml from latest version, and update to include our schema (Note: doing this as opposed to the other way around since it's been so many versions).
2. Run a full reindex on OJF and audit
3. Connect to dev.openlibrary.org and community audit
4. Place in prod environment hooked up to solr-updater connected to dev and audit
5. Replace live solr with new solr.

### Stakeholders
@tfmorris

**Comment by cdrini** [comment]
From current reindex (based on an old dump; I forget which one exactly):

| Type | # in postgres dump | # in solr | diff |
| -- | -- | -- | -- |
| Works | ... | 21,956,868 | ... |
| Orphans | ... | 2,314,615 | ... |
| Authors | ... | 8,357,946 | ... |
| Subjects | ... | ... | ... |

**Comment by cdrini** [comment]
Ok, bit the bullet and just implemented subjects indexing. It actually works super fast! Indexing all the subjects took <5min!

Here are the numbers::

| Subject Type | Count in old Solr | Count in Solr8 |
| ------------ | ----------------- | -------------- |
| subject      | 512,972           | 1,856,864      |
| person       | 536,971           | 741,915        |
| place        | 232,213           | 329,616        |
| time         | 18,373            | 28,648         |
| org\*        | 189,728           | 0              |
| work\*       | 18,512            | 0              |
| event\*      | 5,295             | 0              |
| **Total**    | **1,514,064**     | **2,957,043** |

\* These appear to have been deprecated at some point?

**Comment by cdrini** [comment]
Status update: Tried to run on the latest dump, but the dump was missing records :( Blocked by #4621 .

**Comment by cdrini** [comment]
Status Update: Merged the first batch of PRs (including the jumbo #4937 !). Extracted the next 2 PRs from #4337 .

Started the full reindex on solr1; waiting for it to finish to see if the data dump fixes worked. There's still the bug that `text_general` has converted everything to multiValued (eg author_name is now multiValued). Will see if I can fix and run another full reindex later. Determining the validity of the data dumps while I find a fix seems good.

**Comment by cdrini** [comment]
Completed without a hitch!


Type | # in postgres dump | # in solr | diff
-- | -- | -- | --
Works | 22,127,655 | 22,127,654 | -1
Orphans | 2,310,485 | 2,310,478 | -7
Authors | 8,426,018 | 8,425,773 | -245
Subjects | NA |   | NA

Cause of missing authors is unclear; no errors were logged.

Orphans had 1 error:

```
2021-04-15 13:57:42,952 [ERROR] Error in normalizing 'PA-8161.00000000.M871 1919\x04\x04\x04\x04 .M871 1919'
2021-04-15 13:57:42,952 [ERROR] failed to update work /works/OL17913211M
Traceback (most recent call last):
  File "openlibrary/solr/update_work.py", line 1336, in openlibrary.solr.update_work.update_work
  File "openlibrary/solr/update_work.py", line 739, in openlibrary.solr.update_work.dict2element
  File "openlibrary/solr/update_work.py", line 156, in openlibrary.solr.update_work.add_field_list
  File "openlibrary/solr/update_work.py", line 139, in openlibrary.solr.update_work.add_field
  File "openlibrary/solr/update_work.py", line 136, in openlibrary.solr.update_work.add_field
  File "src/lxml/etree.pyx", line 1024, in lxml.etree._Element.text.__set__
  File "src/lxml/apihelpers.pxi", line 747, in lxml.etree._setNodeText
  File "src/lxml/apihelpers.pxi", line 735, in lxml.etree._createTextNode
  File "src/lxml/apihelpers.pxi", line 1540, in lxml.etree._utf8
```

And many warnings about author redirects.


Works had 1 error, many warning about missing title/author redirects.

```
2021-04-15 09:39:09,989 [ERROR] failed to update work /works/OL24233388W
Traceback (most recent call last):
  File "openlibrary/solr/update_work.py", line 1336, in openlibrary.solr.update_work.update_work
  File "openlibrary/solr/update_work.py", line 739, in openlibrary.solr.update_work.dict2element
  File "openlibrary/solr/update_work.py", line 156, in openlibrary.solr.update_work.add_field_list
  File "openlibrary/solr/update_work.py", line 139, in openlibrary.solr.update_work.add_field
  File "openlibrary/solr/update_work.py", line 136, in openlibrary.solr.update_work.add_field
  File "src/lxml/etree.pyx", line 1024, in lxml.etree._Element.text.__set__
  File "src/lxml/apihelpers.pxi", line 747, in lxml.etree._setNodeText
  File "src/lxml/apihelpers.pxi", line 735, in lxml.etree._createTextNode
  File "src/lxml/apihelpers.pxi", line 1540, in lxml.etree._utf8
ValueError: All strings must be XML compatible: Unicode or ASCII, no NULL bytes or control characters
```

[works-indexing.log.gz](https://github.com/internetarchive/openlibrary/files/6321676/works-indexing.log.gz)
[orphans-indexing.log.gz](https://github.com/internetarchive/openlibrary/files/6321677/orphans-indexing.log.gz)
[authors-indexing.log.gz](https://github.com/internetarchive/openlibrary/files/6321678/authors-indexing.log.gz)

Calling that a success!

**Comment by cdrini** [comment]
Next steps:

- [ ] Spin up a new solr-updater that points to ol-solr1 so we can see how it handles updates
- [ ] Point staging to ol-solr1
- [ ] Fix bug with `text_general`

**Comment by cdrini** [comment]
Note I uploaded the data to IA, so if you want to try loading it on your computer, see https://archive.org/details/olsolr8-2021-04-12 . You'll need ~25GB of hard disk space :+1:

**Comment by cdrini** [comment]
Status update: code for solr-updater mods merged (Namely #5056, #5101); Cleanup to some solrbuilder code merged: #5048 . Away tomorrow, but planning to do a deploy Wendesday which will put the solr-updater bound to solr8 up and running!

**Comment by cdrini** [comment]
2 above done! New bug found:

- [x] solr8-updater throwing error about content-type missing
- [ ] All search results sometimes empty?

**Comment by cdrini** [comment]
Drats! Some nasty bugs :(

- [ ] Search results are sometimes empty because it's handling stop words weirdly. A query that contains stop words (like "of" or "the") seems to no longer match. eg http://testing.openlibrary.org/search?q=%22mark+of+the+crown%22&mode=everything vs http://testing.openlibrary.org/search?q=title%3Amark+crown&mode=everything
- [ ] This syntax seems to not work anymore for some reason: `key:(/works/OL450777W OR /works/OL362289W)`. Need to do `key:("/works/OL450777W" OR "/works/OL362289W")`

**Comment by cdrini** [comment]
Ok, figured out the stop words issue. Using the solr analysis tool, we see:

http://localhost:8983/solr/#/openlibrary/analysis?analysis.fieldvalue=The%20Mark%20of%20the%20Crown&analysis.fieldname=title&verbose_output=0&analysis.query=the%20mark%20of%20the%20crown

Note that the last rows are not the same!

![image](https://user-images.githubusercontent.com/6251786/117372418-7ae42980-ae97-11eb-83f7-d712ed9e7181.png)

Looking at the schema, the `title` field has `fieldType` `text_en_splitting`. It looks like this field flattens adjacent "skips" during indexing, but not during querying! Which is what's causing the mismatch. Adding the `FGF` (`FlattenGraphFilter`) to query analysis fixes it :+1: Although `all

| Query                           | Before (OL UI) | After Query + FGF (OL UI) |
| ------------------------------- | --- | --- |
| `title:The Mark of the Crown`   | ✅  | ✅ |
| `title:"The Mark of the Crown"` | ❌  | ✅ |
| `"The Mark of the Crown"`       | ❌  | ✅ |
| `The Mark of the Crown`         | ❌  | ❌ |

Although when querying solr directly:

| Query                           | Before (Solr direct) | After Query + FGF (Solr direct) |
| ------------------------------- | --- | --- |
| `title:The Mark of the Crown`   | ✅  | ✅ |
| `title:"The Mark of the Crown"` | ❌  | ✅ |
| `"The Mark of the Crown"`       | ❌  | ✅ |
| `The Mark of the Crown`         | ✅  | ✅ |

So it looks like the remaining issue is to do with the way our `all` search works.


I feel a little 👎 about modifying the default `text_en_splitting`? Honestly this feel like it might be a bug in the default. I email Solr's mailing list to see if I should open an issue on their JIRA: See https://lists.apache.org/thread.html/r6fc3beb0c2e9f37d7dead7894ab1270ecbb28153368d259e651bce57%40%3Cusers.solr.apache.org%3E

**Comment by cdrini** [comment]
Next problem: all search not working with `mark of the crown`, but working for `mark crown`

This is the request made to solr:

```
q=mark+of+the+crown
&fq=type:work
&facet=true
&facet.field=has_fulltext
&facet.field=author_facet
&facet.field=language
&facet.field=first_publish_year
&facet.field=publisher_facet
&facet.field=subject_facet
&facet.field=person_facet
&facet.field=place_facet
&facet.field=time_facet
&facet.field=public_scan_b
&fl=key,author_name,author_key,title,subtitle,edition_count,ia,has_fulltext,first_publish_year,cover_i,cover_edition_key,public_scan_b,lending_edition_s,lending_identifier_s,language,ia_collection_s
&rows=20
&start=0
&q.op=AND
&spellcheck=true
&spellcheck.count=3

# dismax is making it fail :/
&defType=dismax
&qf=text+title^5+author_name^5
&bf=sqrt(edition_count)^10
```

The issue is because `title` and `text` now have stopwords! Here's the parsed query in solr 3.6:

```
+((DisjunctionMaxQuery((author_name:the^5.0 | text:the | title:the^50.0)) DisjunctionMaxQuery((author_name:mark^5.0 | text:mark | title:mark^50.0)) DisjunctionMaxQuery((author_name:of^5.0 | text:of | title:of^50.0)) DisjunctionMaxQuery((author_name:the^5.0 | text:the | title:the^50.0)) DisjunctionMaxQuery((author_name:crown^5.0 | text:crown | title:crown^50.0)))~5) ()
``` ( from http://server.openjournal.foundation:8984/solr/select?q=the%20mark%20of%20the%20crown&fq=type:work&facet=true&facet.field=has_fulltext&facet.field=author_facet&facet.field=language&facet.field=first_publish_year&facet.field=publisher_facet&facet.field=subject_facet&facet.field=person_facet&facet.field=place_facet&facet.field=time_facet&facet.field=public_scan_b&facet.mincount=1&fl=key,author_name,author_key,title,subtitle,edition_count,ia,has_fulltext,first_publish_year,cover_i,cover_edition_key,public_scan_b,lending_edition_s,lending_identifier_s,language,ia_collection_s&rows=20&start=0&q.op=AND&spellcheck=true&spellcheck.count=3&defType=dismax&qf=text+title%5E50+author_name%5E5&debugQuery=on )

And here it is in 8.8:

```
+(DisjunctionMaxQuery(((author_name:the)^5.0)) DisjunctionMaxQuery(((author_name:mark)^5.0 | text:mark | (title:mark)^50.0)) DisjunctionMaxQuery(((author_name:of)^5.0)) DisjunctionMaxQuery(((author_name:the)^5.0)) DisjunctionMaxQuery(((author_name:crown)^5.0 | text:crown | (title:crown)^50.0)))~5 ()
``` ( via http://localhost:8983/solr/openlibrary/select?q=the%20mark%20of%20the%20crown&fq=type:work&facet=true&facet.field=has_fulltext&facet.field=author_facet&facet.field=language&facet.field=first_publish_year&facet.field=publisher_facet&facet.field=subject_facet&facet.field=person_facet&facet.field=place_facet&facet.field=time_facet&facet.field=public_scan_b&facet.mincount=1&fl=key,author_name,author_key,title,subtitle,edition_count,ia,has_fulltext,first_publish_year,cover_i,cover_edition_key,public_scan_b,lending_edition_s,lending_identifier_s,language,ia_collection_s&rows=20&start=0&q.op=AND&spellcheck=true&spellcheck.count=3&defType=dismax&qf=text+title^50+author_name^5&debugQuery=on )



Note that because `the` is never going to match `title`/`text` since it's a stop word, instead of `(author_name:the^5.0 | text:the | title:the^50.0)` we have `(author_name:the)^5.0`! So now it becomes a mandatory match on `author_name` :/

**Comment by cdrini** [comment]
Ok, deciding to remove stopword detection for now (our solr 3.6 didn't have stopwords). We can figure it out at a later date. Without stopword:

Query | Before | After Disabling stopwrads
-- | -- | --
title:The Mark of the Crown | ✅ | ✅
title:"The Mark of the Crown" | ❌ | ✅
"The Mark of the Crown" | ❌ | ✅
The Mark of the Crown | ❌ | ✅

**Comment by cdrini** [comment]
Gaaah! `key:/works/OL19446266W` isn't working because it's being parsed as a RegExp 😱 This was added in Solr 4, and queries with `/.../` are treated as regex :(

Solr 8:
http://localhost:8983/solr/openlibrary/select?q=key:/works/OL19446266W&debugQuery=on
```xml
<str name="parsedquery">
RegexpQuery(key:/works/) (text:ol text:19446266 text:w)
</str>
<str name="parsedquery_toString">key:/works/ (text:ol text:19446266 text:w)</str>
```

**Comment by cdrini** [comment]
The multiValued bug turned out to be easy; turns out I can add `multiValued="false"` and it doesn't complain! 🎉

**Comment by cdrini** [comment]
Aaaand, just escaped the `/` from the python for the regex issue :P #5192 .

Just kicked off another (hopefully final before production deploy!) full reindex!

**Comment by cdrini** [comment]
Grumble. After a bunch of failure and a ton of fixes to solrbuilder, finally got a successful run: 


Type | # in postgres dump | # in solr | diff
-- | -- | -- | --
Works | 22,264,179 | 22,264,178 | -1
Orphans | 2,307,941 | 2,307,934 | -7
Authors | 8,487,789 | 8,487,543 | -246
Subjects | NA | 2,957,272 | NA

Looking good. Starting off the "catchup"

**Comment by cdrini** [comment]
Reading for another batch of testing on http://testing.openlibrary.org !

**Comment by cdrini** [comment]
testing.openlibrary.org went well (1 fix: #5209); deployed to one of our web nodes yesterday. Performance/load times seem to be looking good! Edits are still showing up within <2min in search results \o/ . Going to try deploying to both our web nodes to see if the new solr can handle that level of traffic.

**Comment by cdrini** [comment]
Ok, perf issues! Here are my potential suspects

(1) Our solr-updater triggers a solr `<commit/>` every minute; this seems to cause a noticeable dip in performance. With solr8, we should be making use of automatic soft commits, and not explicitly calling commit. That will require some testing.
   - Still happening after stopped solr-updater...

(2) ~Confirm this server is using an SSD; can't recall if it is.~ Confirmed it is.

(3) Need to run a solr optimize? Docs say might not be needed with modern solr, but worth trying
    - Tried hitting /solr/openlibrary/update?optimize=true and it took like 5s? That's definitely not right; should take ~30min.
    - Looks like it needs to be /solr/openlibrary/update?optimize=true&maxSegments=1 (Not that that's documented anywhere except stack overflow :/ )

(4) Need an haproxy? The old solr sat behind an haproxy; was hoping I wouldn't need one for solr8...

**Comment by cdrini** [comment]
Optimize complete; down to 1 segment (took 50min). Hooking it all back up again!

**Comment by cdrini** [comment]
Things I've tried:

- [x] Starting solr8-update with `--no-commit`. It doesn't need to commit, since it's using the `commitWithin` flag set to 60s :/
    - Confirmed edits still going through ~every minute
- [x] Disabled replication in the UI - nothing changed as far as I can tell?
- [x] Confirmed SSD

Going to try adding an haproxy layer.

**Comment by cdrini** [comment]
Ok, solr (behind haproxy) hooked to production! Let's see how that does.

Also I've been monitoring the number of segments, and it's stabilized ~5-8 now? Before the optimize it was 26. So the optimize seems like a good thing to keep after the full reindex.

**Comment by cdrini** [comment]
Woohoo! Performance seems to be peachy now :) Although `--no-commit` seems a little suspicious; it feels like `commitWithin` isn't always being respected? Will leave it on for a while...

**Comment by cdrini** [comment]
Ok, that made things immediately worse :/ Taking back off.

**Comment by cdrini** [comment]
Status update; running on prod healthily! Will clean up/split the PR one last time, and then on track for merging :)

**Comment by cdrini** [comment]
Status update; first of final two PRs merged (#5320 ), the big ol' #4337 !
